49-tgyuuAn

tgyuuAn opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[KCM Travel](https://www.acmicpc.net/problem/10217)

**문제가 쉬워보여서 선택.**

**그리고 큰 코 뿌러짐**

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
**4시간?**

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/116813010/b997f69e-5b4f-4b26-b973-f59b219d59d1)

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/116813010/af68c508-a812-4691-a1cf-e31b49374e1e)

- 실제로는 생각하고 푼 것이라 소모 시간이 훨씬 많음.
- Python이나 pypy3으로 했을 때 다익스트라로는 풀리지 않는다. DP로 접근해야함.

<br><br><br><br><br><br>

처음에는 **다익스트라**를 이용해야하나 ? 싶었다.

다익스트라를 이용해서 **2차원 DP 테이블**을 만든 뒤 `DP[N][C] = T` 에서 N번째 노드에 C 코스트를 들였을 때 얻을 수 있는 최소 시간을 저장해나갔다.

이렇게 만드니 답은 잘 나왔다.

하지만, 아무리 노력해도 파이썬으로 **시간 초과**를 면할 수가 없었다.

```python
from collections import defaultdict, deque
import sys

def input(): return sys.stdin.readline().rstrip()

T = int(input())
for _ in range(T):
    N, M, K = map(int, input().split())

    # i 번째 인덱스 항의 값에서 1. 비용 2. 시간 3. 인덱스
    costs = [[int(1e9) for _ in range(M+1)] for x in range(N+1)]
    board = deque([(0, 0, 1)])
    costs[1][1] = 0

    graph = [[] for _ in range(N+1)]
    for _ in range(K):
        start, destination, cost, duration = map(int,input().split())
        graph[start].append((destination, cost, duration))

    # print(edges)

    while board:
        now_cost, now_duration, now_idx = board.popleft()
        
        if now_duration > costs[now_idx][now_cost]: continue
    
        for destination, cost, duration in graph[now_idx]:
            if now_cost + cost > M: continue

            if costs[destination][now_cost + cost] > now_duration + duration:
                costs[destination][now_cost + cost] = now_duration + duration
                
                # 도착지점이 목표면 삽입할 필요 없음.                
                if destination == N: continue

                board.append((now_cost + cost, now_duration + duration, destination))

    result = min(costs[N])
    print(result) if result != int(1e9) else print("Poor KCM")
```

<br><br><br><br><br><br>

음... 도저히 위 코드에서 시간을 줄일 수 있는 방법을 떠올리지 못해서 열심히 서칭을 진행했다.

<br><br><br><br><br><br>

**허거걱...**

[다익스트라를 쓰지 않고 DP만을 이용해서 푸는 방법](https://velog.io/@ledcost/%EB%B0%B1%EC%A4%80-10217-%ED%8C%8C%EC%9D%B4%EC%8D%AC-KCM-Travel-%EA%B3%A8%EB%93%9C-1-DP)이 존재했다.

<br><br><br><br><br><br>

기존에 가방에 물건채우기 알고리즘인 냅색과 유사하게 아래와 같이 2중 반복문을 돌리면 해결할 수 있었다.

이 때 중요한 점은 가장 외곽이 코스트여야 하고 내부에 city가 있어야 한다는 점이다.

(city가 외곽에 있을 경우 다익스트라 처럼 주변노드만을 먼저 탐색하는 것이 아니라,

갈 수 없는 지점에서 탐색하는 경우가 생길 수 있다. 

예를 들어 1번 노드에서 2번 노드를 가지 못하는데, 

2번 노드의 탐색을 시작함으로써 해당 노드로부터 가는 티켓은 모두 무효가_ 되는 것처럼 말이다.)

```python
for cost in range(M+1):
    for city in range(1, N+1):
        # ... 코드 작성~
```

<br><br><br><br><br><br>

PR은 되게 짧은데 여기 까지 오는데 엄청난 시행 착오를 겪었다...

PR로 쓰고나니 되게 허무하네..

이 문제 약간 [앱](https://github.com/AlgoLeadMe/AlgoLeadMe-1/pull/139) 문제의 상위 호환 인 것 같아요..!

되게 어렵게 풀었는데 이렇게 밖에 설명을 못하겠따...

ㅋㅌㅋㅌㅋㅋㅋ... 파이팅!!!!!

**파이썬 다익스트라로 풀면 밑에 써줘요!!**

## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

9kyo-hwang commented:
이거 질문게시판 보니까 8달 전 쯤에 다익스트라 저격 테케 추가돼서 싹 다 날라갔더라;;
그래서 구글링해서 나오는 코드들 대부분 다 지금 통과 안되는 코드들임 ㅋㅋ

9kyo-hwang reviewed (approved):
C++로 통과한 다익스트라 코드 중 제일 빠른 거 변환하니까 겨우 통과하네요 ㅋㅋ
```py
from heapq import *
input = open(0).readline

for _ in range(T := int(input())):
    N, M, K = map(int, input().split())
    adj_list = [[] for _ in range(N + 1)]
    distances = [[float('inf')] * (M + 1) for _ in range(N + 1)]
    distances[1] = [0] * (M + 1)

    for _ in range(K):
        u, v, c, d = map(int, input().split())
        adj_list[u].append((v, c, d))
        
    def dijkstra():
        heap = [(0, 1)]
        while heap:
            du, u = heappop(heap)
            if distances[u][M] < du:
                continue
            
            for v, cv, dv in adj_list[u]:
                prev_dist = distances[v][M]
                for cost in range(0, M + 1):
                    if cost + cv > M:
                        break
                    
                    if distances[u][cost] == float('inf'):
                        continue
                    
                    if distances[u][cost] + dv < distances[v][cost + cv]:
                        distances[v][cost + cv] = distances[u][cost] + dv
                
                if prev_dist > distances[v][M]:
                    heappush(heap, (distances[v][M], v))
                    
        return print('Poor KCM' if distances[N][M] == float('inf') else distances[N][M])
        
    dijkstra()
```
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/49135176/11dcc323-e59e-4239-b6f1-6a95d521c0f6)

이거 문제 의도 자체가 DP인것 같습니다....

tgyuuAn commented:
> C++로 통과한 다익스트라 코드 중 제일 빠른 거 변환하니까 겨우 통과하네요 ㅋㅋ
> 
> ```python
> from heapq import *
> input = open(0).readline
> 
> for _ in range(T := int(input())):
>     N, M, K = map(int, input().split())
>     adj_list = [[] for _ in range(N + 1)]
>     distances = [[float('inf')] * (M + 1) for _ in range(N + 1)]
>     distances[1] = [0] * (M + 1)
> 
>     for _ in range(K):
>         u, v, c, d = map(int, input().split())
>         adj_list[u].append((v, c, d))
>         
>     def dijkstra():
>         heap = [(0, 1)]
>         while heap:
>             du, u = heappop(heap)
>             if distances[u][M] < du:
>                 continue
>             
>             for v, cv, dv in adj_list[u]:
>                 prev_dist = distances[v][M]
>                 for cost in range(0, M + 1):
>                     if cost + cv > M:
>                         break
>                     
>                     if distances[u][cost] == float('inf'):
>                         continue
>                     
>                     if distances[u][cost] + dv < distances[v][cost + cv]:
>                         distances[v][cost + cv] = distances[u][cost] + dv
>                 
>                 if prev_dist > distances[v][M]:
>                     heappush(heap, (distances[v][M], v))
>                     
>         return print('Poor KCM' if distances[N][M] == float('inf') else distances[N][M])
>         
>     dijkstra()
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/49135176/318273803-11dcc323-e59e-4239-b6f1-6a95d521c0f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE4OTM2NjgsIm5iZiI6MTcxMTg5MzM2OCwicGF0aCI6Ii80OTEzNTE3Ni8zMTgyNzM4MDMtMTFkY2MzMjMtZTU5ZS00MjM5LWI2ZjEtNmE5NWQ1MjFjMGY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzMxVDEzNTYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiNmM0N2U0MGUyYWI2MDVmZmViNmU4NDlhMzMzMDcwNjBhYjM2NmQyYThmMzE1Yjg0YzZmYjdjYzJmNjY1ZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5ggnRls0Fwo2ajxKWGaHpHs72kZsuczqx2Sg7DPHcVo)
> 
> 이거 문제 의도 자체가 DP인것 같습니다....

헐 C++로 시간 뭐임... 파이썬으로 절대 안되겠다...

H0ngJu commented on tgyuuAn/DP/KCM Travel.py:
```diff
@@ -0,0 +1,30 @@
+for _ in range(T):
+    N, M, K = map(int, input().split())
+
+    costs = [[int(1e9) for _ in range(M+1)] for _ in range(N+1)]
```
몇 글자 차이 안나긴 하지만,, INF = int(1e9)로 따로 두는게 어떨까요?
int(1e9)가 자주 사용되니 INF로 따로 상수로 빼두면 가독성이 더 좋아질 것 같습니다 😎

tgyuuAn replied:
헉 완전 동의합니다....!  😁😁😁

H0ngJu reviewed (approved):
ㅈ.. 제 코는 아직 부러질 코도 아니기 때문에 이 문제는 킵해두겠습니다 .. ㅎㅎ
~~플레티넘 보고 도망가는거 아닙니다 ..~~

이건 PR과는 상관없는 이야기인데 백준에서 결과 뒤에 퍼센트가 원래 뜨나요??
저는 항상 퍼센트 올라가다가 최종적으로는 퍼센트 없이 결과만 뜨던데 ㅇ..어떻게 보나요..?!

tgyuuAn commented:
> ㅈ.. 제 코는 아직 부러질 코도 아니기 때문에 이 문제는 킵해두겠습니다 .. ㅎㅎ
> 
> ~~플레티넘 보고 도망가는거 아닙니다 ..~~
> 
> 
> 
> 이건 PR과는 상관없는 이야기인데 백준에서 결과 뒤에 퍼센트가 원래 뜨나요??
> 
> 저는 항상 퍼센트 올라가다가 최종적으로는 퍼센트 없이 결과만 뜨던데 ㅇ..어떻게 보나요..?!

어 저도 이거 준성님한테 배웠는데 크롬에 있는 백준 익스텐션 깔아야해요!! @pknujsp 혹시 이름이 뭐였쬬...?

9kyo-hwang commented:
> > ㅈ.. 제 코는 아직 부러질 코도 아니기 때문에 이 문제는 킵해두겠습니다 .. ㅎㅎ
> > ~플레티넘 보고 도망가는거 아닙니다 ..~
> > 이건 PR과는 상관없는 이야기인데 백준에서 결과 뒤에 퍼센트가 원래 뜨나요??
> > 저는 항상 퍼센트 올라가다가 최종적으로는 퍼센트 없이 결과만 뜨던데 ㅇ..어떻게 보나요..?!
> 
> 어 저도 이거 준성님한테 배웠는데 크롬에 있는 백준 익스텐션 깔아야해요!! @pknujsp 혹시 이름이 뭐였쬬...?

[baekjoonhub](https://chromewebstore.google.com/detail/%EB%B0%B1%EC%A4%80%ED%97%88%EB%B8%8Cbaekjoonhub/ccammcjdkpgjmcpijpahlehmapgmphmk?hl=ko) 쓰시면 됩니다.

H0ngJu commented:
> > > ㅈ.. 제 코는 아직 부러질 코도 아니기 때문에 이 문제는 킵해두겠습니다 .. ㅎㅎ
> > > ~플레티넘 보고 도망가는거 아닙니다 ..~
> > > 이건 PR과는 상관없는 이야기인데 백준에서 결과 뒤에 퍼센트가 원래 뜨나요??
> > > 저는 항상 퍼센트 올라가다가 최종적으로는 퍼센트 없이 결과만 뜨던데 ㅇ..어떻게 보나요..?!
> > 
> > 
> > 어 저도 이거 준성님한테 배웠는데 크롬에 있는 백준 익스텐션 깔아야해요!! @pknujsp 혹시 이름이 뭐였쬬...?
> 
> [baekjoonhub](https://chromewebstore.google.com/detail/%EB%B0%B1%EC%A4%80%ED%97%88%EB%B8%8Cbaekjoonhub/ccammcjdkpgjmcpijpahlehmapgmphmk?hl=ko) 쓰시면 됩니다.

감사합니다.. (_ _) 맨날 퍼센트 올라가는거 쳐다보고 있었는데 따로 있었군요 .. 😅
